src/state.rs

```diff
@@ -1,171 +1,123 @@
 // Copyright (c) DUSK NETWORK. All rights reserved.
 // Licensed under the MPL 2.0 license. See LICENSE file in the project root for details.
 
-use std::io;
 use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
 
-use dataview::Pod;
-use dusk_abi::H256;
-use kelvin::{ByteHash, Content, Sink, Source, ValRef, ValRefMut};
-use kelvin_radix::DefaultRadixMap as RadixMap;
+use canonical::{ByteSource, Canon, Ident, Store};
+use canonical_derive::Canon;
+use dusk_abi::{Query, Transaction};
+use dusk_kelvin_map::Map;
 
 use crate::call_context::{CallContext, Resolver};
-use crate::contract::{Contract, MeteredContract};
+use crate::contract::{Contract, ContractId};
 use crate::gas::GasMeter;
 use crate::VMError;
 
-pub type Storage<H> = RadixMap<Vec<u8>, Vec<u8>, H>;
-
-#[derive(Clone)]
-pub struct ContractState<H: ByteHash> {
-    balance: u128,
-    code: MeteredContract,
-    nonce: u64,
-    storage: Storage<H>,
-}
-
-impl<H: ByteHash> ContractState<H> {
-    pub fn empty() -> Self {
-        ContractState {
-            code: MeteredContract::empty(),
-            balance: 0,
-            nonce: 0,
-            storage: Default::default(),
-        }
-    }
-
-    pub fn balance(&self) -> u128 {
-        self.balance
-    }
-
-    pub fn balance_mut(&mut self) -> &mut u128 {
-        &mut self.balance
-    }
-
-    pub fn storage(&self) -> &Storage<H> {
-        &self.storage
-    }
-
-    pub fn storage_mut(&mut self) -> &mut Storage<H> {
-        &mut self.storage
-    }
-
-    pub fn contract(&self) -> &MeteredContract {
-        &self.code
-    }
-
-    pub fn contract_mut(&mut self) -> &mut MeteredContract {
-        &mut self.code
-    }
-}
-
 /// The main network state, includes the full state of contracts.
-#[derive(Clone, Default)]
-pub struct NetworkState<S, H: ByteHash> {
-    contracts: RadixMap<H256, ContractState<H>, H>,
-    _marker: PhantomData<S>,
+#[derive(Clone, Default, Canon)]
+pub struct NetworkState<E, S>
+where
+    S: Store,
+{
+    contracts: Map<ContractId, Contract, S, 17>,
```

Comment:
Zedger trees will not have `17` depth. We need to think on a solution for trees of different size (maybe upgrade the const definition of the size to the network state itself?)

Also, by definition, contracts may have smaller poseidon trees, if they opt to

Reply:
This will be resolved by the removal of const arrayvec in microkelvin.